core/services/mongodb-sync/src/test/java/com/mongodb/stitch/core/services/mongodb/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -0,0 +1,151 @@
+package com.mongodb.stitch.core.services.mongodb.sync.internal
+
+import com.mongodb.MongoNamespace
+import com.mongodb.stitch.core.StitchAppClientInfo
+import com.mongodb.stitch.core.internal .common.AuthMonitor
+import com.mongodb.stitch.core.internal .net.NetworkMonitor
+import com.mongodb.stitch.core.services.internal .CoreStitchServiceClient
+import com.mongodb.stitch.core.services.mongodb.remote.internal.CoreRemoteMongoClientImpl
+import com.mongodb.stitch.core.services.mongodb.sync.ChangeEventListener
+import com.mongodb.stitch.core.services.mongodb.sync.SyncConflictResolver
+import junit.framework.Assert.*
+import org.bson.*
+import org.bson.codecs.BsonDocumentCodec
+
+import org.bson.codecs.configuration.CodecRegistries
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.mockito.ArgumentCaptor
+import org.mockito.Mockito
+import org.mockito.Mockito.verify
+
+class DataSynchronizerUnitTests {
+    private val testDb = "testDb"
+    private val testColl = "testColl"
+
+    private val namespace = MongoNamespace(testDb, testColl)
+
+    private var dataSynchronizer: DataSynchronizer? = null
+    private var insertedId: BsonValue? = null
+
+    @Before
+    fun setup() {
+        val networkMonitor = object : NetworkMonitor {
+            override fun isConnected(): Boolean {
+                return true
+            }
+
+            override fun addNetworkStateListener(listener: NetworkMonitor.StateListener) {
+
+            }
+        }
+        val authMonitor = AuthMonitor { true }
+        dataSynchronizer = DataSynchronizer(
+                "testInstanceKey",
+                Mockito.mock(CoreStitchServiceClient::class.java),
+                LocalClientFactory.getClient(
+                        StitchAppClientInfo(
+                                "<client-app-id>",
+                                "__store__",
+                                "<local-app-name>",
+                                "<local-app-version>", null,
+                                networkMonitor,
+                                authMonitor
+                        ),
+                        "local"
+                ),
+                Mockito.spy(CoreRemoteMongoClientImpl(Mockito.mock(CoreStitchServiceClient::class.java))),
+                networkMonitor,
+                authMonitor
+        )
+    }
+
+    @After
+    fun teardown() {
+        val dataSpy = Mockito.spy(dataSynchronizer!!)
+        dataSpy.deleteOneById(namespace, insertedId)
+    }
+
+    @Test
+    fun testInsertOneAndSync() {
+        val dataSpy = Mockito.spy(dataSynchronizer!!)
+
+        insertOneAndSync(dataSpy, BsonDocument())
+
+        assertEquals(dataSpy.getSynchronizedDocuments(namespace).first().documentId, insertedId)
+        assertEquals(dataSpy.getSynchronizedDocuments(namespace).size, 1)
```

Comment:
If you make an assertion like this, then you should assert prior to the insert that it's size is 0. I would also check that size is 1 before doing first, otherwise first may not have anything.

Reply:
I would also fill out the BsonDocument, and do a find one on it to verify what you inserted is there. In addition to this, this test doesn't feel complete. The data synchronizer will eventually try to push this to the remote. You could check this by running a sync pass and mock out the insertion for the remote. If you do this, you prove that the change event in the config is of type INSERT (which could also check but this would make it unnecessary)